### DISCUSSION TITLE
Parsing issues with ${{ }}

### DISCUSSION THREAD

**Discussion by Fatty911** [community/community#59062] [Actions] [answered]
### Select Topic Area

Question

### Body

https://github.com/Fatty911/AutoBuild_OpenWRT_for_Mi_R4/actions/runs/5367828756/jobs/9738284012

Please see the execution results of my Actions process, test3 test4 test5 have been executed, test6 has not been executed. 
Please check my workfile, compare my test3 test4 test5 test6, the conditions of the judgment statement are
test3    if: ${{ steps.compile.outputs.status }} == 'success' && env.UPLOAD_BIN_DIR == 'true';
test4    if: steps.compile.outputs.status == 'success' && ${{ env.UPLOAD_BIN_DIR }} == 'true';
test5    if: ${{ steps.compile.outputs.status }} == 'success' && ${{ env.UPLOAD_BIN_DIR }} == 'true';
test6    if: steps.compile.outputs.status == 'success' && env.UPLOAD_BIN_DIR == 'true'.

So why test6 is not executed and other tests can be executed ？

**Comment by airtower-luna** [comment]
The conditions in 3 to 5 will *always* evaluate to `true`, no matter the values you put in. Only test6 actually checks the condition.

The reason: If there's any `${{ ... }}` in your condition, only that part is evaluated as an expression, the rest is treated as a literal string. And a non-empty string always evaluates to `true`. You have to either wrap *everything* in `${{ ... }}`, or nothing (in that case the evaluation code will add it implicitly).

**Reply by Fatty911** [reply]
According to your statement, the test6 expression will add ${{ ... }} implicitly, so that test6 and test5 are the same, but why is test5 executed but test6 not executed?

**Reply by ChristopherHX** [reply]
test5 != test6

test6 can also be written like
```yaml
if: ${{ steps.compile.outputs.status == 'success' && env.UPLOAD_BIN_DIR  == 'true' }}
```

test5 can also be written like (I wonder why this implicit conversion is not reported as an syntax error for if expressions)
```yaml
if: ${{ format('{0} == ''success'' && {1} == ''true''', steps.compile.outputs.status, env.UPLOAD_BIN_DIR) }}
```
Since this string is never empty and contains no status check function => it is always true if `success()` is true.

test6 won't run, because
```bash
echo "{status}={success}" >> $GITHUB_OUTPUT
```
creates an output with name `{status}` not `status` and with value `{success}` and not `success`.

the following
```bash
echo "status=success" >> $GITHUB_OUTPUT
```
would make test6 run.

**Reply by airtower-luna** [reply, marked as answer]
They won't be the same. The implicit evaluation covers *everything*. After adding the implicit `${{ ... }}`, test6 will look like this:

```
${{ steps.compile.outputs.status == 'success' && env.UPLOAD_BIN_DIR == 'true' }}
```

And *only* then are the operators actually evaluated. In all your other examples (outside the `${{ ... }}`) they are only strings that happen to look like operators.

In test5 the values don't matter at all. Let's assume both the output and the environment variable you're looking at are `false`. Then the result of substituting them will be (including escapes):

```
'''false'' == ''success'' && ''false'' == ''true'''
```

But that's a *string* value. And a non-empty string is always `true`.

**Reply by Fatty911** [reply]
thanks !!!

**Reply by Fatty911** [reply]
thanks !!!
